### ISSUE TITLE
The IsInteractable function in InteractionReceiver is not matching  gameobjects from ObjectCollection

### ISSUE THREAD

**Issue by juansuerogit** [Microsoft/MixedRealityToolkit-Unity#1632]
## Overview
The IsInteractables function in InteractionReceiver is not matching my gameobjects from ObjectCollection 
```
       protected bool Isinteractable(GameObject interactable)
       {
           return (interactables != null && interactables.Contains(interactable));
       }
```

## Expected Behavior
Even though im doing this..   
public class RegisterWithInteractionReceiver : MonoBehaviour {
```
   [SerializeField]
   public GameScreen m_InteractionReceiver;

   [SerializeField]
   public ObjectCollection m_Interactables;

   // Use this for initialization
   void Awake () {

       if (m_Interactables != null)
       {
           if (m_InteractionReceiver != null)
           {
               foreach (CollectionNode node in m_Interactables.NodeList)
               {
                   m_InteractionReceiver.Registerinteractable(node.transform.gameObject);

               }
           }
       }
    }
}
```

## Actual Behavior

When i use the immediate window both interactable and interactable[2]  have the same values..
"Difficulty (UnityEngine.GameObject)"
   base: "Difficulty (UnityEngine.GameObject)"
   activeInHierarchy: true
   activeSelf: true
   gameObject: "Difficulty (UnityEngine.GameObject)"
   isStatic: false
   layer: 5
   scene: {UnityEngine.SceneManagement.Scene}
   tag: "Untagged"
   transform: "Difficulty (UnityEngine.Transform)"

## Steps to reproduce
wire up the RegisterWithInteractionReceiver.cs above with a  InteractionReceiver and ObjectCollection representing some buttons

## Unity Editor Version
2017.2.1.f1

## Mixed Reality Toolkit Release Version
Current version MixedRealityToolkit for Unity 2017.2.1f1

**Comment by fogbridgegit** [comment]
im new to C# .. but is see the different objects have different hashcodes?
 
((object)interactable).GetHashCode()
-1286 
((object)interactables[2]).GetHashCode()
-1152

**Comment by fogbridgegit** [comment]
-1286 all the way back to original RayCast of the FocusManager so maybe something wrong with the object collection


public void UpdateHit(RaycastHit hit, RayStep sourceRay, int rayStepIndex)

**Comment by StephenHodgson** [comment]
Yeah they're not the same objects because the `eventData.selectedObject` comes from Unity's mouse pointer click.  It's better to get the currently focused object directly from the Focus manager with `TryGetFocusedObject(event data)`.

**Comment by fogbridgegit** [comment]
i changed my code and removed the ObjectCollection but still doesnt match up 

public class RegisterWithInteractionReceiver : MonoBehaviour {

    [SerializeField]
    public GameScreen m_InteractionReceiver;

    [SerializeField]
    public GameObject m_ParentOfInteractablesToRegister;

    // Use this for initialization
    void Awake () {

        if (m_ParentOfInteractablesToRegister != null)
        {
            if (m_InteractionReceiver != null)
            {
                foreach (Transform t in m_ParentOfInteractablesToRegister.transform)
                {
                    m_InteractionReceiver.Registerinteractable(t.gameObject);

                }
            }
        }
	}
	

}

**Comment by fogbridgegit** [comment]
thanks for the clarification ...

the thing is im not doing anything.. im beholden to how the holotoolkit does it.   

The Holotoolkit InterationReceiver is registered as a global listener with the Holotoolkit Input Manager.  So when the Unity stuff the Holotoolkit is wrapping hands the Holotoolkit an Input Click event , the Holotoolkit  passes the object from the RayCastHit to the Input Manager which then calls the Interaction receiver from the InputManager via this function

        private static readonly ExecuteEvents.EventFunction<IInputClickHandler> OnInputClickedEventHandler =
            delegate (IInputClickHandler handler, BaseEventData eventData)
            {
                var casted = ExecuteEvents.ValidateEventData<InputClickedEventData>(eventData);
                handler.OnInputClicked(casted);
            };


Then the interaction receiver loops through an array of its interactables and does a list.contain that doesnt match.

I havent done anything.

Are you saying i should change the internals of the holotoolkit?

**Comment by StephenHodgson** [comment]
No. I'm not suggesting that at all.

**Comment by StephenHodgson** [comment]
I'm saying you're looking for an object that doesn't get set unless it's inherited from UnityEngine.UI.Graphics.  that's the only time the event data's selectedObject gets set.

I'm saying that you're checking the wrong object for sameness for the raised event.

**Comment by fogbridgegit** [comment]
oh ok.. but im saying that its not me doing the checking its the holotoolkit doing the checking..  im just giving my butttons to Holotoolkit  InteractionManager

**Comment by fogbridgegit** [comment]
all i did was send in the same objects that clicked on into the Holotoolkit InteractionManagers RegisterInteractable and the Holotoolkit does a List.Contains checking what comes from InputManager against the object i just handed to it.

**Comment by StephenHodgson** [comment]
Oh my apologies, you mean the current interaction input receiver is doing this. Yeah maybe that class needs to he updated, but the core input system shouldn't be changed at this time.

**Comment by fogbridgegit** [comment]
its all good no worries just trying to understand.  So it sounds like im stuck :(

**Comment by StephenHodgson** [comment]
I'll take a look today to see.  I'm doing a refactor anyway so I'll be sure to very it's fixed.

**Comment by StephenHodgson** [comment]
But it answer your question, it makes sense they're hash codes are different because they're from two different event data's.

**Comment by fogbridgegit** [comment]
cool , Im slowly trying to move the MRDL sample app to the current Holotoolkit so the way they do menus there is they consolidate input handling into similar InteractionReceiver and drive a number of menus from that.  Instead of Applying the Handlers on the buttons themselves which would be very messy.  The use InteractionReceiver centralize.

**Comment by StephenHodgson** [comment]
The MRDL stuff has a long way to go.

**Comment by StephenHodgson** [comment]
We're working on it tho.

**Comment by fogbridgegit** [comment]
Its for personal learning purposes.

**Comment by fogbridgegit** [comment]
Dude youre not gonna beleive its all my fault.  i was click on 1 menu but had a twin off the scene somewhere that was the real one sending its objects :(:( #eggOnFaceHere #sorry 

i found out when i started injecting a monobehavior with my own guid before i registered the interactible with InteractionReceiver and it was coming back null from the holotoolkit.

t.gameObject.AddComponent<MRTKID>();

ill crawl back in my hole now

**Comment by StephenHodgson** [comment]
Naw man. There was an issue here. We should have getting the Focused object from focus manager.

**Comment by fogbridgegit** [comment]
i see.  i have another issue inside interactive receiver.
  
ive got two menus that run back to back.   both are child of a parent that runs a single interactivereceiver

when i InputDown on a button I call Difficulty in my start menu, its suppose to hide the startmenu and then show the difficulty menu, and then i press a button on the difficulty and supposed to come back to my startmenu.

That works if i dont place the 2nd menu behind the first.  But when i do put the 2nd menu behind the first, even though its deactivated it still gets an InputDown also

Heres my interactive receiver that has interactables from both menus..

        protected override void InputDown(GameObject obj, InputEventData eventData)
        {
            switch (obj.name)
            {
                case "Difficulty":
                    // switch to difficulty menu
                    startupMenu.SetActive(false);
                    difficultyMenu.SetActive(true);
                    break;

What looks like is happening in laymans terms is the raycast is hitting both?
but how could that be?   
i always use eventData.Use(); like you said.

**Comment by fogbridgegit** [comment]
I ran it on the hololens and no problems.  seems to only happen in the editor with the xbox controller.  ill try it in the emulator.

**Comment by fogbridgegit** [comment]
so its an editor thing.  because on the emulator and hololens device it works fine.  :( :( .   but that kills my ability to debug.   maybe i can fix it

**Comment by fogbridgegit** [comment]
Ah!  it was the same thing from the other day.  I disabled one of the XboxControllerInputSource scripts 
![image](https://user-images.githubusercontent.com/24654459/34921931-d7ef14b8-f956-11e7-8ee6-790e4d5e0c10.png)

**Comment by wiwei** [comment]
Closing issues older than 180 days. If this is still an issue, please reactivate with recent information.
